### PR TITLE
rewrite with operator framework

### PR DESCRIPTION
**Pull request by xavpaice**
Reimplement the charm using the operator framework.

Add basic unit testing.

Add nrpe check for the process.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,57 +1,43 @@
...
+	@echo "Cloning submodules"
+	@git submodule update --init --recursive
+
+lint:
+	@echo "Running flake8"
+	@tox -e lint
+
+test: lint unittest functional
+
+unittest:
```
Comment (ivoks) [inline Makefile:24]: unittest should call submodules first, since it depends on those submodules

**Comment by xavpaice** [comment]
Thanks for the review, and spotting that - I'll make sure we update our template also.  I've added a commit with the change.

**Comment by ivoks** [comment]
Xav, I've merged this, but it's far from done. None of the hooks fire (config-changed) and lldpd.conf file is not the same in every ubuntu version. Please don't update the charm in charmstore until all functionality is achieved.

**Comment by ivoks** [comment]
It seems that everything is fine when charm is deployed in green field, but on upgrades nothing happens:

2020-04-30 10:54:04 INFO juju.worker.uniter.charm bundles.go:77 downloading local:bionic/lldpd-3 from API server
2020-04-30 10:54:04 INFO juju.downloader download.go:111 downloading from local:bionic/lldpd-3
2020-04-30 10:54:04 INFO juju.downloader download.go:94 download complete ("local:bionic/lldpd-3")
2020-04-30 10:54:04 INFO juju.downloader download.go:174 download verified ("local:bionic/lldpd-3")
2020-04-30 10:54:05 INFO juju.worker.uniter resolver.go:130 found queued "upgrade-charm" hook
2020-04-30 10:54:06 INFO juju.worker.uniter.operation runhook.go:141 skipped "upgrade-charm" hook (missing)
2020-04-30 10:54:06 INFO juju.worker.uniter resolver.go:130 found queued "config-changed" hook
2020-04-30 10:54:06 INFO juju.worker.uniter.operation runhook.go:141 skipped "config-changed" hook (missing)

Interesting...

**Comment by xavpaice** [comment]
This is because when upgrading the charm, the symlinks in ./hooks aren't being created correctly  - just the install hook is there.  Raised https://github.com/canonical/operator/issues/256

**Pull request merged**